sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -351,11 +492,43 @@ void ProgramManager::build(RT::PiProgram Program, const string_class &Options,
 
   if (!Opts)
     Opts = Options.c_str();
-  if (PI_CALL_NOCHECK(piProgramBuild)(Program, Devices.size(), Devices.data(),
-                                      Opts, nullptr, nullptr) == PI_SUCCESS)
-    return;
 
-  throw compile_program_error(getProgramBuildLog(Program));
+  std::vector<RT::PiProgram> LinkPrograms;
+  if (LinkDeviceLibs) {
+    getDeviceLibPrograms(Context, Devices, CachedLibPrograms, LinkPrograms);
+  }
+
+  if (LinkPrograms.empty()) {
+    RT::PiResult Error = PI_CALL_NOCHECK(piProgramBuild)(
+        Program.get(), Devices.size(), Devices.data(), Opts, nullptr, nullptr);
+    if (Error != PI_SUCCESS)
+      compile_program_error(getProgramBuildLog(Program.get()));
+    return Program;
+  }
+
+  // Include the main program and compile/link everything together
+  PI_CALL(piProgramCompile)
+  (Program.get(), Devices.size(), Devices.data(), Opts, 0, nullptr, nullptr,
```

Comment:
Please fix formatting here

Reply:
This was formatted by clang-format. Fixed.